currencies/rate_converter.go

```diff
@@ -0,0 +1,178 @@
+func NewRateConverter(
+	httpClient httpClient,
+	syncSourceURL string,
+	fetchingInterval time.Duration,
+	retryFetchingInterval time.Duration,
+) *RateConverter {
+	return NewRateConverterWithNotifier(
+		httpClient,
+		syncSourceURL,
+		fetchingInterval,
+		retryFetchingInterval,
+		nil, // no notifier channel specified, won't send any notifications
+	)
+}
+
+// NewRateConverterWithNotifier returns a new RateConverter
+// it allow to pass an update chan in which the number of ticks will be passed after each tick
+// allowing clients to listen on updates
+// Do not use this method
+func NewRateConverterWithNotifier(
+	httpClient httpClient,
+	syncSourceURL string,
+	fetchingInterval time.Duration,
+	retryFetchingInterval time.Duration,
+	updateNotifier chan<- int,
+) *RateConverter {
+	rc := &RateConverter{
+		httpClient:            httpClient,
+		done:                  make(chan bool),
+		updateNotifier:        updateNotifier,
+		fetchingInterval:      fetchingInterval,
+		retryFetchingInterval: retryFetchingInterval,
+		ticker:                nil,
+		syncSourceURL:         syncSourceURL,
+		rates:                 atomic.Value{},
+		lastUpdated:           atomic.Value{},
+	}
+
+	// In case host do not want to support currency lookup
+	// we just stop here and do nothing
+	if rc.fetchingInterval == time.Duration(0) {
+		return rc
+	}
+
+	rc.Update()                // Make sure to populate data before to return the converter
+	rc.startPeriodicFetching() // Start periodic ticking
+
+	return rc
+}
+
+// fetch allows to retrieve the currencies rates from the syncSourceURL provided
+func (rc *RateConverter) fetch() (*Rates, error) {
+	request, err := http.NewRequest("GET", rc.syncSourceURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	request = request.WithContext(ctx)
+
+	response, err := rc.httpClient.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	bytesJSON, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedRates := &Rates{}
+	err = json.Unmarshal(bytesJSON, updatedRates)
+	if err != nil {
+		return nil, err
+	}
+
+	return updatedRates, err
+}
+
+// Update updates the internal currencies rates from remote sources
+func (rc *RateConverter) Update() error {
+	rates, err := rc.fetch()
+	if err == nil {
+		rc.rates.Store(rates)
+		rc.lastUpdated.Store(time.Now())
+	} else {
+		glog.Errorf("Error updating conversion rates: %v", err)
+	}
+
+	return err
+}
+
+// startPeriodicFetching starts the periodic fetching at the given interval
+// triggers a first fetch when called before the first tick happen in order to initialize currencies rates map
+// returns a chan in which the number of data updates everytime a new update was done
+func (rc *RateConverter) startPeriodicFetching() error {
+
+	rc.ticker = time.NewTicker(rc.fetchingInterval)
+	updatesTicksCount := 0
```

Comment:
There are some race conditions in here... `rc.ticker` is written in this scope, but read from inside the goroutine. In general, shared memory is dangerous.

I think the `ticker` should be a local variable. By saving it on `rc`, race conditions are very likely. `updateTicksCount` should be confined to the goroutine where it's used as well.

A few other suggestions:

1. You're always returning `nil`.... so the return value is useless. You may as well remove it.
2. In idiomatic Go, synchronous methods are the "default." I'm no expert... but I think this function should block forever, and you should run it with `go rc.startPeriodicFetching()`. I'd be curious if @zachbadgett had some insight here.

Reply:
@dbemiller is exactly right. Another thing about this goroutine is that it will cause a CPU spike due to the `default` keyword. The goroutine will no longer "park"/yield to the scheduler. This function should look more like this:
```go
func (rc *RateConverter) startPeriodicFetching(fetchingInterval time.Duration) {
	t := time.NewTicker(fetchingInterval)
	var updatesTicksCount int
	for {
		select {
		case <-t.C:
			// TODO: Handle retry strategy
			rc.Update()
			updatesTicksCount++
			if rc.updateNotifier != nil {
				rc.updateNotifier <- updatesTicksCount
			}
		case <-rc.done:
			t.Stop()
			return
		}
	}
}
```

Reply:
oh man... good catch! I didn't even see that.

Thanks for taking a look!

Reply:
Oh I didn't know about that, thanks guys, it's way simpler :)

Reply:
Done

Reply:
Thanks @hhhjort you are right, I just changed it, thanks